File: src/main/java/com/hf/hfw/api/v1/BudgetService.java
package com.hf.hfw.api.v1;

import com.hf.homefinanceshared.Account;
import com.hf.homefinanceshared.Budget;
import com.hf.homefinanceshared.BudgetCategory;
import java.util.List;
import javax.ws.rs.Consumes;
import javax.ws.rs.DELETE;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;

/**
 *
 * @author pldorrell
 */
public interface BudgetService {
    @GET
    @Produces({"application/json"})
    @Consumes({"application/json"})
    @Path("/search/all")    
    public List<Budget> getBudgets();
    
    @GET
    @Produces({"application/json"})
    @Consumes({"application/json"})
    @Path("/getbyId/{id}")    
    public Budget getById(@PathParam("id") String id);
    
    @POST
    @Produces({"application/json"})
    @Consumes({"application/json"})
    @Path("/save")    
    public Budget saveBudget(Budget budget) throws Exception;
    
    @DELETE
    @Produces({"application/json"})
    @Consumes({"application/json"})
    @Path("/delete/{id}")
    public void deleteBudget(@PathParam("id") String id);
    
    @POST
    @Produces({"application/json"})
    @Consumes({"application/json"})
    @Path("/saveBudgetItem/{id}/{type}/")    
    public Budget saveBudgetItem(@PathParam("id") String id,@PathParam("type") String type,BudgetCategory budgetCategory) throws Exception;

    @DELETE
    @Produces({"application/json"})
    @Consumes({"application/json"})
    @Path("/deleteBudgetItem/{id}/{type}/")    
    public Budget deleteBudgetItem(@PathParam("id") String id,@PathParam("type") String type,BudgetCategory budgetCategory) throws Exception;
}

File: src/main/webapp/script/app/services/budgetservice.js
angular.module('BudgetServiceModule',[]);
angular.module('BudgetServiceModule').service("BudgetService", function ($http) {
    //var users = ["Peter", "Daniel", "Nina"]
    var BudgetService = {};

    BudgetService.getBudgets = function () {
        return $http({
            method: 'GET',
            url: '/services/api/v1/budgets/search/all'
        });
    };
    BudgetService.saveBudget = function (budget) {
        delete budget.$$hashKey;
        angular.forEach(budget.incomeItems, function (value, key) {
            console.log(value);
            delete value.$$hashKey;

            //this.push(key + ': ' + value);
            //if (value.category != '') {
            //    $scope.registryTransactionFormCategorySplits.push(value);    
            //}

        });
        angular.forEach(budget.expenseItems, function (value, key) {
            console.log(value);
            delete value.$$hashKey;

            //this.push(key + ': ' + value);
            //if (value.category != '') {
            //    $scope.registryTransactionFormCategorySplits.push(value);    
            //}

        });
        angular.forEach(budget.transferItems, function (value, key) {
            console.log(value);
            delete value.$$hashKey;

            //this.push(key + ': ' + value);
            //if (value.category != '') {
            //    $scope.registryTransactionFormCategorySplits.push(value);    
            //}

        });
        var csrfHeaderName = $("meta[name='_csrf_header']").attr("content");
        var csrfHeaderValue = $("meta[name='_csrf']").attr("content");
        return $http({
            method: 'POST',
            url: '/services/api/v1/budgets/save',
            data: JSON.stringify(budget), // pass in data as strings
            headers: {'Content-Type': 'application/json', 'X-CSRF-TOKEN': csrfHeaderValue}  // set the headers so angular passing info as form data (not request payload)
        });
    };
    BudgetService.validateTransaction = function (transaction) {
        delete transaction.$$hashKey;
        var csrfHeaderName = $("meta[name='_csrf_header']").attr("content");
        var csrfHeaderValue = $("meta[name='_csrf']").attr("content");
        return $http({
            method: 'POST',
            url: '/services/api/v1/register/validate',
            data: JSON.stringify(transaction), // pass in data as strings
            headers: {'Content-Type': 'application/json', 'X-CSRF-TOKEN': csrfHeaderValue}  // set the headers so angular passing info as form data (not request payload)
        });
    };
    return BudgetService;
});
